src/meili/match_route.cc

```diff
@@ -234,7 +234,12 @@ std::vector<EdgeSegment> ConstructRoute(const MapMatcher& mapmatcher,
         // we modify the first segment of the new_segments accordingly to replace the previous
         // one in the route.
         new_segments.front().source = route.back().source;
-        new_segments.front().first_match_idx = route.back().first_match_idx;
+        // Prefer first_match_idx from previous segments but do not replace valid value with invalid.
+        if (route.back().first_match_idx != -1)
+          new_segments.front().first_match_idx = route.back().first_match_idx;
+        // Prefer last_match_idx from new segments but do not replace valid value with invalid.
+        if (new_segments.front().last_match_idx == -1)
+          new_segments.front().last_match_idx = route.back().last_match_idx;
```

Comment:
i've finally had the time to actually run the code and think about what is going on here. so yeah here we are replacing the last segment in the last portion of the match that we found with the first segment of the next portion of the match we found. since they occur on the same edge we basically need to merge the two segment records. this means we want the first match index to be earliest point we would have seen in the match, ie from `route.back()` but its possible that really only one match point landed on this edge in which case `route.back()` would only have set the `last_match_idx` this is why we have to check for "unset". similarly when setting the `last_match_idx` of the merged record we know that the `new_segments` would have seen the "latest" point that landed on this edge, however if, again only one point would have landed on that edge, it would have only set `first_match_idx`. in this case we can be sure that `route.back()` will have its `last_match_idx` set.

basically when only one point falls on a given edge segment and we need to merge the segment records there, we should set both the first and last match_indices to the same index. i think the fix you have accomplishes that though i cant shake the feeling that it could be simplified.